src/document.jl

```diff
@@ -270,17 +270,83 @@ julia> tokens(sd)
     "."
 ```
 """
-tokens(d::(Union{FileDocument, StringDocument})) = tokenize(language(d), text(d))
+tokens(d::(Union{FileDocument,StringDocument})) = tokenize(language(d), text(d))
 tokens(d::TokenDocument) = d.tokens
 function tokens(d::NGramDocument)
     error("The tokens of an NGramDocument cannot be reconstructed")
 end
 
-tokens!(d::TokenDocument, new_tokens::Vector{T}) where {T <: AbstractString} = (d.tokens = new_tokens)
-function tokens!(d::AbstractDocument, new_tokens::Vector{T}) where T <: AbstractString
+tokens!(d::TokenDocument, new_tokens::Vector{T}) where {T<:AbstractString} = (d.tokens = new_tokens)
+function tokens!(d::AbstractDocument, new_tokens::Vector{T}) where {T<:AbstractString}
     error("The tokens of a $(typeof(d)) cannot be directly edited")
 end
 
+
+##############################################################################
+#
+# vocab() / vocab!(): Access to document text as a vocabulary
+#
+# to_string_vector(): Helper function for creating a vocabulary from a StringDocument or a Vector{String}
+#
+##############################################################################
+# Converts a StringDocument to Vector{String}
+to_string_vector(doc::StringDocument) = tokens(doc)
+# Identity function for Vector{String}
+to_string_vector(vec::Vector{String}) = vec
+
+"""
+    vocab(input::Union{StringDocument, Vector{String}}) -> OrderedDict{String, Int}
+
+Create an ordered dictionary from a `StringDocument` or a `Vector` of strings (useful for creating cooccurrence matrices with coo_matrix() (cf. example below). The dictionary maps each unique string to its corresponding index.
+
+# Arguments
+- `input::Union{StringDocument, Vector{String}}`: Input can be either a `StringDocument` or a `Vector{String}`. 
+  For `StringDocument`, the tokens are extracted and used. For `Vector{String}`, the vector itself is used.
+
+# Returns
+- `OrderedDict{String, Int}`: An ordered dictionary where each key is a unique string from the input, 
+  and the value is the index of that string in the original input.
+
+# Examples
+```julia
+julia> doc = StringDocument("This is a sample sentence of a sample document.");
+       vocab(doc) 
+
+OrderedDict{String, Int64} with 8 entries:
+  "This"     => 1
+  "is"       => 2
+  "a"        => 3
+  "sample"   => 4
+  "sentence" => 5
+  ⋮          => ⋮
+
+julia> str_vec = ["This", "is", "a", "sample", "sentence", "of", "a", "sample", "document"];
+       vocab(str_vec)
+       
+OrderedDict{String, Int64} with 7 entries:
+  "This"     => 1
+  "is"       => 2
+  "a"        => 3
+  "sample"   => 4
+  "sentence" => 5
+  ⋮          => ⋮
+"""
+function vocab(input::Union{StringDocument,Vector{String}})
+    string_vector = to_string_vector(input)
+    string_vector = length(string_vector) != length(unique(string_vector)) ? unique(string_vector) : string_vector
```

Comment:
Why not just use `string_vector = to_string_vector(input) |> unique`? If the array has already unique values, nothing will be changed.  In your implementation there is twice call of the `unique` in the worst case.

Reply:
Right! Much better!